src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+
+        private void AdjustConditionalState(BoundExpression node)
+        {
+            if (IsConstantTrue(node))
             {
-                VisitCondition(catchBlock.ExceptionFilterOpt);
-                SetState(StateWhenTrue);
+                Unsplit();
+                SetConditionalState(this.State, UnreachableState());
+            }
+            else if (IsConstantFalse(node))
+            {
+                Unsplit();
+                SetConditionalState(UnreachableState(), this.State);
+            }
+            else if ((object)node.Type == null || node.Type.SpecialType != SpecialType.System_Boolean)
+            {
+                // a dynamic type or a type with operator true/false
+                Unsplit();
             }
 
-            VisitStatement(catchBlock.Body);
+            Split();
         }
 
-        protected virtual void VisitFinallyBlock(BoundStatement finallyBlock, ref TLocalState unsetInFinally)
+        /// <summary>
+        /// Visit a general expression, where we will only need to determine if variables are
+        /// assigned (or not). That is, we will not be needing AssignedWhenTrue and
+        /// AssignedWhenFalse.
+        /// </summary>
+        /// <param name="node"></param>
+        protected BoundNode VisitRvalue(BoundExpression node)
         {
-            VisitStatement(finallyBlock); // this should generate no pending branches
+            Debug.Assert(!_trackExceptions || this.PendingBranches.Count > 0 && this.PendingBranches[0].Branch == null);
+            var result = Visit(node);
+            Unsplit();
+            return result;
         }
 
-        #endregion TryStatements
+        /// <summary>
+        /// Visit a statement.
+        /// </summary>
+        [DebuggerHidden]
+        protected virtual void VisitStatement(BoundStatement statement)
+        {
+            Debug.Assert(!_trackExceptions || this.PendingBranches.Count > 0 && this.PendingBranches[0].Branch == null);
+            Visit(statement);
+            Debug.Assert(!this.IsConditionalState);
+        }
+
+        protected static bool IsConstantTrue(BoundExpression node)
+        {
+            return node.ConstantValue == ConstantValue.True;
+        }
+
+        protected static bool IsConstantFalse(BoundExpression node)
+        {
+            return node.ConstantValue == ConstantValue.False;
+        }
+
+        protected static bool IsConstantNull(BoundExpression node)
+        {
+            return node.ConstantValue == ConstantValue.Null;
+        }
+
+        /// <summary>
+        /// Called at the point in a loop where the backwards branch would go to.
+        /// </summary>
+        private void LoopHead(BoundLoopStatement node)
+        {
+            TLocalState previousState;
+            if (_loopHeadState.TryGetValue(node, out previousState))
+            {
+                Join(ref this.State, ref previousState);
+            }
+
+            _loopHeadState[node] = this.State.Clone();
```

Comment:
could use comment.  i have no clear sense of whne .Clone is needed and when it isn't.